incremental determinization

chenzhehuai opened the pull request:
the main modification is after the function GetLattice() in lattice-incremental-decoder.cc

Commit:
WIP

Commit:
worse wer & ower

Commit:
clean code

chenzhehuai commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+  for (IterType iter = token2pos.begin(); iter != token2pos.end(); ++iter)
+    (*topsorted_list)[iter->second] = iter->first;
+}
+
```
compared to lattice-faster-decoder.cc, the main modification is after this line.

chenzhehuai commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+      int32 id = state_label_map_.find(tok)->second; // it should exist
+      // TODO: calculate alpha but not use tot_cost
+      BaseFloat cost_offset = tok->tot_cost;
+      state_label_forward_prob_[id] = tok->tot_cost;
```
apply cost offset using alpha (tot_cost)

danpovey replied:
You should label it state_label_forward_cost_ if it is a cost rather than a prob (they have opposite sign).

chenzhehuai commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+
+          CompactLatticeWeight weight_offset;
+          weight_offset.SetWeight(
+              LatticeWeight(0, -state_label_forward_prob_[arc_append.olabel]));
```
cancel cost offset using alpha (tot_cost) obtained from the map

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+    for (Token *tok = active_toks_[frame_begin].toks; tok != NULL; tok = tok->next) {
+      StateId cur_state = tok_map[tok];
+      // state_label_map_ is construct during create_final_state
+      int32 id = state_label_map_.find(tok)->second; // it should exist
```
you should probably issue an error if not found rather than just segfaulting.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+  typedef Arc::StateId StateId;
+  typedef Arc::Weight Weight;
+  typedef Arc::Label Label;
+
```
This code is super unstructured; it's not a good practice to write code like this.  Better to break it into functions, classes, etc., with clear interfaces, and document things clearly as you go..  This may be part of why you can't find the bug.

chenzhehuai replied:
This function is mostly the same to GetRawLattice(), I add explanation and comments in the code to describe the difference

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,448 @@
+      state_label_forward_prob_; // alpha for each state_label (Token)
+  // specific design for incremental GetLattice
+  bool GetRawLattice(Lattice *ofst, bool use_final_probs, int32 frame_begin,
+                     int32 frame_end, bool create_initial_state,
```
I don't see how create_initial_state == false would make sense.  And you should document things like this as you write them.

chenzhehuai replied:
we do not create the initial state in the first chunk
we do not create the final state in the last chunk
comments added

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+    for (Token *tok = active_toks_[f].toks; tok != NULL; tok = tok->next) {
+      StateId cur_state = tok_map[tok];
+      for (ForwardLinkT *l = tok->links; l != NULL; l = l->next) {
+        if (f == frame_begin && create_initial_state && l->ilabel == 0)
```
This code is not well structured.  It' s very hard for me to verify that it's correct because it is not written in a clear way. comments might fix this though.

danpovey reviewed (commented):
I can't figure out what's wrong because this code is too unclear.
Even if it had worked, you would have needed to rewrite this almost from scratch.
Why not just do it right the first time?

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+        if (r != initial_arc_map.end()) {
+          ArcIterator<CompactLattice> aiter_chunk(clat, 0); // initial state
+          aiter_chunk.Seek(r->second);
+          const auto &arc_chunk = aiter_chunk.Value();
```
I don't see how 'chunk' in the name helps clarify anything.  You say in the comment that there are two chunks.  So which one is it?

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+    if (last_get_lattice_frame_ != 0)
+      state_offset--; // since we do not append initial state
+    unordered_map<int32, size_t>
+        initial_arc_map; // the incoming states of these arcs are initial states
```
document what this maps from and to.  There needs to be *much* more documentation here.

danpovey replied:
and your comment is not clear.  "the incoming states of these arcs are initial states".  Initial states in what?  Initial states under what definition?  In an FST?  Or in some other sense?

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+              arc_append_mod.weight, arc_chunk.weight, olat->Final(prev_final_state),
+              weight_offset};
+          BaseFloat v1 = 0, v2 = 0;
+          for (auto &i : weights) v1 += i.Weight().Value1();
```
you could have just used Times() on the weights to multiply them, it would have been simpler, if slightly less efficient.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+          if (prev_final_state == -1)
+            prev_final_state = arc_append.nextstate;
+          else
+            KALDI_ASSERT(arc_append.nextstate == prev_final_state);
```
You probably shouldn't rely on the determinized lattice having a single final state.  It will due to an interaction between the specific structure of this RawLattice and how the determinization algorithm works, but this is not something that is super obvious.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+          continue; // will exist in the next chunk
+        typename unordered_map<Token *, StateId>::const_iterator iter =
+            tok_map.find(l->next_tok);
+        StateId nextstate = iter->second;
```
swap with the next line, or could segfault

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+        ofst->AddArc(cur_state, arc);
+      }
+      if (f == frame_end) {
+        if (use_final_probs && !final_costs.empty()) {
```
This does not look right to me, and it might be the reason why your WERs are bad (could cause it to delete later parts of some utterances).  It looks to me like you are setting final-probs in the raw lattice... you should instead only be making transitions to a single final-state.

danpovey replied:
... oh I see, you are deleting them later.  But that is a very unclear way to structure the code-- don't do it that way!

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+      int32 id = state_label_avilable_idx_++;
+      state_label_map_[tok] = id;
+      Weight final_weight =
+          (!decoding_finalized_ && ofst->Final(cur_state) == Weight::Zero())
```
This doesn't look like the right logic for deciding the final weights (plus it's an extremely unclear expression; I suspect you meant || instead of &&, but  that wouldn't fix it.
The logic should be more similar to around line 1145.  If we are using final-probs (e.g. it's the last frame of decoding and the user specified use_final_probs) *and * at least one active state was final, then you should use the final probs for all states, even if most were zero.

danpovey replied:
... and use an if-statement rather than a boolean expression like this, it will  be clearer (I think you even confused yourself here).   This might be the reason why your WERs are not right, it will give the chunks semi-random final probs.

chenzhehuai replied:
Thanks. I will re-work on this part. 
Actually, these boolean expression is of no use, since the value of ofst->Final(cur_state) has been taken care in line 1145.  we can just make 
Weight final_weight = ofst->Final(cur_state);

Ya, I should merge these two parts and set final weight and states in one place.

chenzhehuai replied:
Thanks. I will re-work on this part. 
Actually, these boolean expression is of no use, since the value of ofst->Final(cur_state) has been taken care in line 1145.  we can just do:
Weight final_weight = ofst->Final(cur_state);

Ya, I should merge these two parts and set the final weight and states in one place.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+    state_label_map_.reserve(std::min((int32)1e5, config_.max_active));
+    for (Token *tok = active_toks_[frame_end].toks; tok != NULL; tok = tok->next) {
+      StateId cur_state = tok_map[tok];
+      int32 id = state_label_avilable_idx_++;
```
avilable is misspelled.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1199 @@
+    for (size_t i = 0; i < token_list.size(); i++)
+      if (token_list[i] != NULL) tok_map[token_list[i]] = ofst->AddState();
+  }
+  // The next statement sets the start state of the output FST.  Because we
```
this comment is no longer valid in this context.

Commit:
this commit is for sanity check

Commit:
code clean

Commit:
each time we determinize the piece of lattice, instead of going all the way to the currently-decoded frame, we go up to, say, t-10 (unless this is the end of the utterance), and the same way that we put in temporary initial-probs, we also put in temporary final-probs which reflect the  on the states at frame t-10.  (we remove them later on, of course).

Commit:
bug fix:
1. in the determinized lattice, there could be multiple final arcs with the same state label. I need to change the logic here.
2. for the first chunk, there could be some final arcs starting from state 0, while for the last chunk, there could be some initial arcs ending in final state. Hence, I found that we cannot distinguish final and initial arcs by simply "if (s==0)" or "if (clat.Final(arc_appended.nextstate)!=CompactLatticeWeight::Zero()"

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+    // step 1: Get lattice chunk with initial state
+    // In this function, we do not create the initial state in
+    // the first chunk, and we do not create the final state in the last chunk
+    KALDI_ASSERT(GetRawLattice(&raw_fst, use_final_probs, last_get_lattice_frame_,
```
it's dangerous to put statements with side effects in KALDI_ASSERT, as code can be compiled with -DNDEBUG

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+  if (not_first_chunk)
+    state_offset--; // since we do not append initial state in the first chunk
+
+  // A map between state label and the arc position (index)
```
map FROM x TO x.  'between' applies a transitive relationship.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+  // A map between state label and the arc position (index)
+  // the incoming states of these arcs are initial states of the chunk
+  // and the olabel of these arcs are the key of this map (state label)
+  // The arc position are obtained from ArcIterator corresponding to the state
```
this is a map from what to pair of (what and what)?  You never clearly state this.

danpovey replied:
forget it, I now see it's not a pair.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+    for (auto &i : final_arc_list_prev_) {
+      MutableArcIterator<CompactLattice> aiter(olat, i.first);
+      aiter.Seek(i.second);
+      // Obtain the appended final arcs in the last chunk
```
the previous chunk.  `last` is ambiguous.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+      // of this chunk. If not, it is pruned by PruneActiveTokens()
+      auto r = initial_arc_map.find(arc_appended.olabel);
+      if (r != initial_arc_map.end()) {
+        ArcIterator<CompactLattice> aiter_unappended(clat, 0); // initial state
```
unappeneded is not very helpful, IMO.  Maybe you could call the two arcs that you are going to append arc1 and arc2 and then have just 'arc_appended' be the appended one.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+        weight_offset_final.SetWeight(
+            LatticeWeight(0, -state_label_final_cost_[arc_appended.olabel]));
+        arc_appended_mod.weight =
+            Times(arc_appended_mod.weight, arc_unappended.weight);
```
Probably `Times(a, Times(b, Times(c, d))` would be clearer.. could be split into multiple lines if needed.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+
+        CompactLatticeWeight weight_offset, weight_offset_final;
+        weight_offset.SetWeight(
+            LatticeWeight(0, -state_label_initial_cost_[arc_appended.olabel]));
```
it is bad practice to look up things in maps without checking that they are there first.  This might be masking a bug, see below.

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+        aiter.SetValue(arc_appended_mod);
+      } // otherwise, it has been pruned
+      state_label_initial_cost_.erase(arc_appended.olabel);
+      state_label_final_cost_.erase(arc_appended.olabel);
```
This is dangerous, because can you guarantee that there will be just a single final-arc per symbol?  I don't think you can.  You might be deleting these costs that will later be needed.  And anyway, why would you delete them one by one if you don't need to?  It's wasteful and unnecessary.

chenzhehuai replied:
Yes, I found that in the determinized lattice, there could be multiple final arcs with the same state label. I need to change the logic here.

chenzhehuai commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+      }
+      // Process state labels, which will be used in step 3.2
+      if (arc.olabel > config_.max_word_id) { // initial_arc
+        if (s == 0) { // record initial_arc in this chunk, we will use it right now
```
@danpovey  I found that for the first chunk, there could be some final arcs starting from state 0, while for the last chunk, there could be some initial arcs ending in final state. Hence, I found that we cannot distinguish final and initial arcs by simply "if (s==0)" or "if (clat.Final(arc_appended.nextstate)!=CompactLatticeWeight::Zero()"
How about we make the state label of a initial arc = the corresponding state label of its final arc + a certain offset.
By this way we can explicitly distinguish final and initial arcs in clat here

danpovey replied:
Hm.  Don't you know, in this part of the code, whether this chunk is initial and whether it is final (meaning: whether it is the first chunk, and whether it is the last chunk)?

You could use logic like:
```
  if (s == 0 && !is_initial_chunk) {
     // process initial arc (meaning: arc at the beginning of a chunk with a state label on it
  } else {
    // process final arc (meaning: arc at the end of the chunk with a state label on it).
  }
```   
that would avoid the hassle of messing around with states.

chenzhehuai replied:
Sure I can do something like this if you prefer

chenzhehuai commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1295 @@
+    // step 2: Determinize the chunk
+    CompactLattice clat;
+#if 1
+    if (!DeterminizeLatticePhonePrunedWrapper(
```
@danpovey Another problem I found is that only if we use a beam larger than beam+lattice_beam here, we can guarantee no final or initial arcs are pruned by this function (I already use extra_cost and tot_cost as final and initial arc weights respectively). 
These pruned final arcs can significantly hurt oracle WER performance but not 1-best WER (also result in less lattice density).
Moreover, if I use beam here, the oracle WER performance can be similar to the baseline (also similar lattice density).
Hence, do you prefer using (lattice_beam+beam) or (lattice_beam) or (beam) here?

In summary, after fixing these 3 problems (this problem,  "same state label multiple final arcs" and "distinguish final and initial arcs" I discuss before this conversation), we can get the same 1-best hypotheses in a testset subset.

danpovey commented:
Cool.

I am hoping that in practice we can use just 'beam' as long as you have a
short delay, like the 25 frames we spoke of, or maybe 10 frames.  It may
prune some states that we don't want to prune, but anyway pruned Viterbi
search is approximate.  Or we could use beam + lattice_beam if the extra
time taken in determinization is not significant.

I do expect this approach will be slower than determizing once at the end
of the utterance.   What I'm hoping is we can get it acceptably fast.
There are online applications where you have to be able to generate the
lattices as you go.  I will also try to think of ways to solve this problem
more elegantly.

Can you please work on making the code well structured and documented?





On Fri, Mar 29, 2019 at 11:36 PM Zhehuai Chen <notifications@github.com>
wrote:

> *@chenzhehuai* commented on this pull request.
> ------------------------------
>
> In src/decoder/lattice-incremental-decoder.cc
> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r270611774>:
>
> > +
> +  // last_get_lattice_frame_ is used to record the first frame of the chunk
> +  // last time we obtain from calling this function. If it reaches NumFramesDecoded()
> +  // we cannot generate any more chunk
> +  if (last_get_lattice_frame_ < last_frame_of_chunk) {
> +    Lattice raw_fst;
> +    // step 1: Get lattice chunk with initial state
> +    // In this function, we do not create the initial state in
> +    // the first chunk, and we do not create the final state in the last chunk
> +    KALDI_ASSERT(GetRawLattice(&raw_fst, use_final_probs, last_get_lattice_frame_,
> +                               last_frame_of_chunk, not_first_chunk,
> +                               !decoding_finalized_));
> +    // step 2: Determinize the chunk
> +    CompactLattice clat;
> +#if 1
> +    if (!DeterminizeLatticePhonePrunedWrapper(
>
> @danpovey <https://github.com/danpovey> Another problem I found is that
> only if we use a beam larger than beam+lattice_beam here, we can guarantee
> no final or initial arcs are pruned by this function (I already use
> extra_cost and tot_cost as final and initial arc weights respectively).
> These pruned final arcs can significantly hurt oracle WER performance but
> not 1-best WER (also result in less lattice density).
> Moreover, if I use beam here, the oracle WER performance can be similar to
> the baseline (also similar lattice density).
> Hence, do you prefer using (lattice_beam+beam) or (lattice_beam) or (beam)
> here?
>
> In summary, after fixing these 3 problems (this problem, "same state label
> multiple final arcs" and "distinguish final and initial arcs" I discuss
> before this conversation), we can get the same 1-best hypotheses in a
> testset subset.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chenzhehuai/kaldi/pull/28#pullrequestreview-220806047>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuzhPuULOxk46v-d7kf-0e2VmMXZ5ks5vbtvLgaJpZM4cN3KY>
> .
>

Commit:
test in libri speech
+ grep -H Overall exp_dec/incre.fl.1f/base/ora.base exp_dec/incre.fl.1f/base/ora.den.base exp_dec/incre.fl.1f/incre/ora.base exp_dec/incre.fl.1f/incre/ora.den.base
exp_dec/incre.fl.1f/base/ora.base:LOG (lattice-oracle[5.5.276~4-6f366]:main():lattice-oracle.cc:383) Overall %WER 1.70591 [ 342 / 20048, 109 insertions, 22 deletions, 211 substitutions ]
exp_dec/incre.fl.1f/base/ora.den.base:LOG (lattice-depth[5.5.276~4-6f366]:main():lattice-depth.cc:79) Overall density is 25.1613 over 244027 frames.
exp_dec/incre.fl.1f/incre/ora.base:LOG (lattice-oracle[5.5.276~4-6f366]:main():lattice-oracle.cc:383) Overall %WER 1.80567 [ 362 / 20048, 108 insertions, 25 deletions, 229 substitutions ]
exp_dec/incre.fl.1f/incre/ora.den.base:LOG (lattice-depth[5.5.276~4-6f366]:main():lattice-depth.cc:79) Overall density is 28.1682 over 244027 frames.
+ grep -H WER exp_dec/incre.fl.1f/base/wer exp_dec/incre.fl.1f/incre/wer
exp_dec/incre.fl.1f/base/wer:%WER 12.57 [ 2532 / 20138, 305 ins, 287 del, 1940 sub ]
exp_dec/incre.fl.1f/incre/wer:%WER 12.57 [ 2532 / 20138, 305 ins, 287 del, 1940 sub ]
+ grep real exp_dec/incre.fl.1f/base/log/decode.1.log exp_dec/incre.fl.1f/incre/log/decode.1.log
exp_dec/incre.fl.1f/base/log/decode.1.log:LOG (latgen-faster-mapped[5.5.276~4-6f366]:main():latgen-faster-mapped.cc:164) Time taken 48.4324s: real-time factor assuming 100 frames/sec is 0.912442
exp_dec/incre.fl.1f/incre/log/decode.1.log:LOG (latgen-incremental-mapped[5.5.276~4-6f366]:main():latgen-incremental-mapped.cc:164) Time taken 54.6669s: real-time factor assuming 100 frames/sec is 1.0299

Commit:
clean; without class

chenzhehuai commented:
> Cool. I am hoping that in practice we can use just 'beam' as long as you have a short delay, like the 25 frames we spoke of, or maybe 10 frames. It may prune some states that we don't want to prune, but anyway pruned Viterbi search is approximate. Or we could use beam + lattice_beam if the extra time taken in determinization is not significant. I do expect this approach will be slower than determizing once at the end of the utterance. What I'm hoping is we can get it acceptably fast. There are online applications where you have to be able to generate the lattices as you go. I will also try to think of ways to solve this problem more elegantly. Can you please work on making the code well structured and documented?
> […](#)
> On Fri, Mar 29, 2019 at 11:36 PM Zhehuai Chen ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In src/decoder/lattice-incremental-decoder.cc <[#28 (comment)](https://github.com/chenzhehuai/kaldi/pull/28#discussion_r270611774)>: > + + // last_get_lattice_frame_ is used to record the first frame of the chunk + // last time we obtain from calling this function. If it reaches NumFramesDecoded() + // we cannot generate any more chunk + if (last_get_lattice_frame_ < last_frame_of_chunk) { + Lattice raw_fst; + // step 1: Get lattice chunk with initial state + // In this function, we do not create the initial state in + // the first chunk, and we do not create the final state in the last chunk + KALDI_ASSERT(GetRawLattice(&raw_fst, use_final_probs, last_get_lattice_frame_, + last_frame_of_chunk, not_first_chunk, + !decoding_finalized_)); + // step 2: Determinize the chunk + CompactLattice clat; +#if 1 + if (!DeterminizeLatticePhonePrunedWrapper( @danpovey <https://github.com/danpovey> Another problem I found is that only if we use a beam larger than beam+lattice_beam here, we can guarantee no final or initial arcs are pruned by this function (I already use extra_cost and tot_cost as final and initial arc weights respectively). These pruned final arcs can significantly hurt oracle WER performance but not 1-best WER (also result in less lattice density). Moreover, if I use beam here, the oracle WER performance can be similar to the baseline (also similar lattice density). Hence, do you prefer using (lattice_beam+beam) or (lattice_beam) or (beam) here? In summary, after fixing these 3 problems (this problem, "same state label multiple final arcs" and "distinguish final and initial arcs" I discuss before this conversation), we can get the same 1-best hypotheses in a testset subset. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#28 (review)](https://github.com/chenzhehuai/kaldi/pull/28#pullrequestreview-220806047)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ADJVuzhPuULOxk46v-d7kf-0e2VmMXZ5ks5vbtvLgaJpZM4cN3KY> .

chenzhehuai commented:
> Cool. I am hoping that in practice we can use just 'beam' as long as you have a short delay, like the 25 frames we spoke of, or maybe 10 frames. It may prune some states that we don't want to prune, but anyway pruned Viterbi search is approximate. Or we could use beam + lattice_beam if the extra time taken in determinization is not significant. I do expect this approach will be slower than determizing once at the end of the utterance. What I'm hoping is we can get it acceptably fast. There are online applications where you have to be able to generate the lattices as you go. I will also try to think of ways to solve this problem more elegantly. Can you please work on making the code well structured and documented?
> […](#)
> On Fri, Mar 29, 2019 at 11:36 PM Zhehuai Chen ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In src/decoder/lattice-incremental-decoder.cc <[#28 (comment)](https://github.com/chenzhehuai/kaldi/pull/28#discussion_r270611774)>: > + + // last_get_lattice_frame_ is used to record the first frame of the chunk + // last time we obtain from calling this function. If it reaches NumFramesDecoded() + // we cannot generate any more chunk + if (last_get_lattice_frame_ < last_frame_of_chunk) { + Lattice raw_fst; + // step 1: Get lattice chunk with initial state + // In this function, we do not create the initial state in + // the first chunk, and we do not create the final state in the last chunk + KALDI_ASSERT(GetRawLattice(&raw_fst, use_final_probs, last_get_lattice_frame_, + last_frame_of_chunk, not_first_chunk, + !decoding_finalized_)); + // step 2: Determinize the chunk + CompactLattice clat; +#if 1 + if (!DeterminizeLatticePhonePrunedWrapper( @danpovey <https://github.com/danpovey> Another problem I found is that only if we use a beam larger than beam+lattice_beam here, we can guarantee no final or initial arcs are pruned by this function (I already use extra_cost and tot_cost as final and initial arc weights respectively). These pruned final arcs can significantly hurt oracle WER performance but not 1-best WER (also result in less lattice density). Moreover, if I use beam here, the oracle WER performance can be similar to the baseline (also similar lattice density). Hence, do you prefer using (lattice_beam+beam) or (lattice_beam) or (beam) here? In summary, after fixing these 3 problems (this problem, "same state label multiple final arcs" and "distinguish final and initial arcs" I discuss before this conversation), we can get the same 1-best hypotheses in a testset subset. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#28 (review)](https://github.com/chenzhehuai/kaldi/pull/28#pullrequestreview-220806047)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ADJVuzhPuULOxk46v-d7kf-0e2VmMXZ5ks5vbtvLgaJpZM4cN3KY> .

Do you think we need to move step 1-4 all into a private class? If so, we need to pass active_toks_ and final_costs_ to the private class for GetRawLattice. Since we create initial and final arcs in GetRawLattice, it is highly related to the incremental determinizer.

Commit:
add class LatticeIncrementalDeterminizer

danpovey commented:
Cool. I am hoping that in practice we can use just 'beam' as long as you
> have a short delay, like the 25 frames we spoke of, or maybe 10 frames. It
> may prune some states that we don't want to prune, but anyway pruned
> Viterbi search is approximate. Or we could use beam + lattice_beam if the
> extra time taken in determinization is not significant. I do expect this
> approach will be slower than determizing once at the end of the utterance.
> What I'm hoping is we can get it acceptably fast. There are online
> applications where you have to be able to generate the lattices as you go.
> I will also try to think of ways to solve this problem more elegantly. Can
> you please work on making the code well structured and documented?
> … <#m_4635420944884900238_>
> On Fri, Mar 29, 2019 at 11:36 PM Zhehuai Chen ***@***.***> wrote: ***@.*
> commented on this pull request. ------------------------------ In
> src/decoder/lattice-incremental-decoder.cc <#28 (comment)
> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r270611774>>: >
> + + // last_get_lattice_frame_ is used to record the first frame of the
> chunk + // last time we obtain from calling this function. If it reaches
> NumFramesDecoded() + // we cannot generate any more chunk + if
> (last_get_lattice_frame_ < last_frame_of_chunk) { + Lattice raw_fst; + //
> step 1: Get lattice chunk with initial state + // In this function, we do
> not create the initial state in + // the first chunk, and we do not create
> the final state in the last chunk + KALDI_ASSERT(GetRawLattice(&raw_fst,
> use_final_probs, last_get_lattice_frame_, + last_frame_of_chunk,
> not_first_chunk, + !decoding_finalized_)); + // step 2: Determinize the
> chunk + CompactLattice clat; +#if 1 + if
> (!DeterminizeLatticePhonePrunedWrapper( @danpovey
> <https://github.com/danpovey> https://github.com/danpovey Another problem
> I found is that only if we use a beam larger than beam+lattice_beam here,
> we can guarantee no final or initial arcs are pruned by this function (I
> already use extra_cost and tot_cost as final and initial arc weights
> respectively). These pruned final arcs can significantly hurt oracle WER
> performance but not 1-best WER (also result in less lattice density).
> Moreover, if I use beam here, the oracle WER performance can be similar to
> the baseline (also similar lattice density). Hence, do you prefer using
> (lattice_beam+beam) or (lattice_beam) or (beam) here? In summary, after
> fixing these 3 problems (this problem, "same state label multiple final
> arcs" and "distinguish final and initial arcs" I discuss before this
> conversation), we can get the same 1-best hypotheses in a testset subset. —
> You are receiving this because you were mentioned. Reply to this email
> directly, view it on GitHub <#28 (review)
> <https://github.com/chenzhehuai/kaldi/pull/28#pullrequestreview-220806047>>,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ADJVuzhPuULOxk46v-d7kf-0e2VmMXZ5ks5vbtvLgaJpZM4cN3KY
> .
>
> Do you think we need to move step 1-4 all into a private class? If so, we
> need to pass active_toks_ and final_costs_ to the private class for
> GetRawLattice. Since we create initial and final arcs in GetRawLattice, it
> is highly related to the incremental determinizer.
>
I'm OK with passing those things in.  The point of making it a separate
class is to separate thing as much as possible and clearly document the
code.

It would be a class declared outside the decoder, which the decoder would
have as a member.

danpovey commented:
Think about various options for how to structure it and the best way to
break it up into functions that are easy to document.

When I code I spend 20% of the time thinking about algorithmic issues,
since they are usually relatively straightforward, and 80% of the time
thinking about how to make the code clear, well structured and
self-explanatory.


On Sat, Mar 30, 2019 at 9:14 PM Daniel Povey <dpovey@gmail.com> wrote:

>
>
> Cool. I am hoping that in practice we can use just 'beam' as long as you
>> have a short delay, like the 25 frames we spoke of, or maybe 10 frames. It
>> may prune some states that we don't want to prune, but anyway pruned
>> Viterbi search is approximate. Or we could use beam + lattice_beam if the
>> extra time taken in determinization is not significant. I do expect this
>> approach will be slower than determizing once at the end of the utterance.
>> What I'm hoping is we can get it acceptably fast. There are online
>> applications where you have to be able to generate the lattices as you go.
>> I will also try to think of ways to solve this problem more elegantly. Can
>> you please work on making the code well structured and documented?
>> … <#m_-6185348949243686675_m_4635420944884900238_>
>> On Fri, Mar 29, 2019 at 11:36 PM Zhehuai Chen ***@***.***> wrote: ***@.*
>> commented on this pull request. ------------------------------ In
>> src/decoder/lattice-incremental-decoder.cc <#28 (comment)
>> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r270611774>>: >
>> + + // last_get_lattice_frame_ is used to record the first frame of the
>> chunk + // last time we obtain from calling this function. If it reaches
>> NumFramesDecoded() + // we cannot generate any more chunk + if
>> (last_get_lattice_frame_ < last_frame_of_chunk) { + Lattice raw_fst; + //
>> step 1: Get lattice chunk with initial state + // In this function, we do
>> not create the initial state in + // the first chunk, and we do not create
>> the final state in the last chunk + KALDI_ASSERT(GetRawLattice(&raw_fst,
>> use_final_probs, last_get_lattice_frame_, + last_frame_of_chunk,
>> not_first_chunk, + !decoding_finalized_)); + // step 2: Determinize the
>> chunk + CompactLattice clat; +#if 1 + if
>> (!DeterminizeLatticePhonePrunedWrapper( @danpovey
>> <https://github.com/danpovey> https://github.com/danpovey Another
>> problem I found is that only if we use a beam larger than beam+lattice_beam
>> here, we can guarantee no final or initial arcs are pruned by this function
>> (I already use extra_cost and tot_cost as final and initial arc weights
>> respectively). These pruned final arcs can significantly hurt oracle WER
>> performance but not 1-best WER (also result in less lattice density).
>> Moreover, if I use beam here, the oracle WER performance can be similar to
>> the baseline (also similar lattice density). Hence, do you prefer using
>> (lattice_beam+beam) or (lattice_beam) or (beam) here? In summary, after
>> fixing these 3 problems (this problem, "same state label multiple final
>> arcs" and "distinguish final and initial arcs" I discuss before this
>> conversation), we can get the same 1-best hypotheses in a testset subset. —
>> You are receiving this because you were mentioned. Reply to this email
>> directly, view it on GitHub <#28 (review)
>> <https://github.com/chenzhehuai/kaldi/pull/28#pullrequestreview-220806047>>,
>> or mute the thread
>> https://github.com/notifications/unsubscribe-auth/ADJVuzhPuULOxk46v-d7kf-0e2VmMXZ5ks5vbtvLgaJpZM4cN3KY
>> .
>>
>> Do you think we need to move step 1-4 all into a private class? If so, we
>> need to pass active_toks_ and final_costs_ to the private class for
>> GetRawLattice. Since we create initial and final arcs in GetRawLattice, it
>> is highly related to the incremental determinizer.
>>
> I'm OK with passing those things in.  The point of making it a separate
> class is to separate thing as much as possible and clearly document the
> code.
>
> It would be a class declared outside the decoder, which the decoder would
> have as a member.
>
>

chenzhehuai commented:
> I'm OK with passing those things in. The point of making it a separate class is to separate thing as much as possible and clearly document the code. It would be a class declared outside the decoder, which the decoder would have as a member.

We need to pass std::vector\<TokenList\> active_toks_ to the determinizer class to generate raw lattice. TokenList is a protected class of decoder class, hence I was thinking how about making determinizer class protected

danpovey commented:
Either make the decoder class a friend class of the determinizer class, or
template the determinizer class on the TokenList type.

On Sun, Mar 31, 2019 at 10:17 AM Zhehuai Chen <notifications@github.com>
wrote:

> I'm OK with passing those things in. The point of making it a separate
> class is to separate thing as much as possible and clearly document the
> code. It would be a class declared outside the decoder, which the decoder
> would have as a member.
>
> We need to pass std::vector active_toks_ to the determinizer class to
> generate raw lattice. TokenList is a protected class of decoder class,
> hence I was thinking how about making determinizer class protected
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chenzhehuai/kaldi/pull/28#issuecomment-478345946>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu8JQ0SsFWaRjapCXELQFawLPt5uYks5vcMOKgaJpZM4cN3KY>
> .
>

chenzhehuai commented:
> Either make the decoder class a friend class of the determinizer class, or template the determinizer class on the TokenList type.

GetRawLattice needs TopSortTokens() and ComputeFinalCosts() previous defined in decoder class. Do you think we need to move them into determinizer class if we want to do GetRawLattice  in determinizer? Notably, ComputeFinalCosts() is also needed by many functions in decoder class.
Otherwise, do you think GetRawLattice with creating initial and final arcs (with state labels) in decoder class, and then passing the raw lattice to determinizer will be better?

danpovey commented:
Maybe the determinizer class can just be responsible for stitching together
the pieces generated by GetRawLattice and determinizing them.


On Sun, Mar 31, 2019 at 7:28 PM Zhehuai Chen <notifications@github.com>
wrote:

> Either make the decoder class a friend class of the determinizer class, or
> template the determinizer class on the TokenList type.
>
> GetRawLattice needs TopSortTokens() and ComputeFinalCosts() previous
> defined in decoder class. Do you think we need to move them into
> determinizer class if we want to do GetRawLattice in determinizer? Notably,
> ComputeFinalCosts() is also needed by many functions in decoder class.
> Otherwise, do you think GetRawLattice with creating initial and final arcs
> (with state labels) in decoder class, and then passing the raw lattice to
> determinizer will be better?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chenzhehuai/kaldi/pull/28#issuecomment-478395659>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVuylXhzYHdez84IqUygmfU-IEqxxxks5vcUSGgaJpZM4cN3KY>
> .
>

Commit:
add config_.determinize_max_active & redeterminize=false

Commit:
update best config; add re-determinization from frame 0 if AppendLatticeChunks failed

Commit:
1. add time profiling for baseline lattice-faster-decoder for comparison (remove it later)
2. add determinize-beam-offset. By this way, the beam used in lattice determinization
is (determinize_beam_offset + lattice_beam)

Commit:
code refine

Commit:
update final weight by extra_cost-alpha, see sheet "ver 3"

Commit:
WIP

Commit:
fix bugs and add sanity check

Commit:
enable det

Commit:
clean code

chenzhehuai commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,602 @@
+  /// and figures out the shortest path.
+  bool GetBestPath(Lattice *ofst, bool use_final_probs = true);
+
+  // The following function is specifically designed for incremental
```
try to describe what we do as a whole here. Notably, I separate previous  `state_label` to two terms:  `state_label` and `token_label`

chenzhehuai commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,602 @@
+  // Output the resultant determinized lattice in the form of CompactLattice
+  const CompactLattice &GetDeterminizedLattice() const { return lat_; }
+
+  // Part of step 1 of incremental determinization,
```
the new initial arc generation

chenzhehuai commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1423 @@
+      // of final arcs to denote the "future" information of final states (Tokens)
+      int32 frame_det_least = last_get_lattice_frame_ + config_.prune_interval;
+      if (config_.determinize_lattice && frame_det_most > 0) {
+        // To adaptively decide the length of chunk, we further compare the number of
```
the adaptive delay is here

chenzhehuai reviewed (commented):
@danpovey  Could you please check this version and give some comments on the coding style and documents. Let me know the part mainly need to be improved

danpovey commented on src/bin/Makefile:
```diff
@@ -17,6 +17,7 @@ BINFILES = align-equal align-equal-compiled acc-tree-stats \
         post-to-weights sum-tree-stats weight-post post-to-tacc copy-matrix \
         copy-vector copy-int-vector sum-post sum-matrices draw-tree \
         align-mapped align-compiled-mapped latgen-faster-mapped latgen-faster-mapped-parallel \
+				latgen-incremental-mapped \
```
strange whitespace here.  You should probably add this at the end of the list of programs.

danpovey commented on src/bin/latgen-incremental-mapped.cc:
```diff
@@ -0,0 +1,179 @@
+// Copyright 2009-2012  Microsoft Corporation, Karel Vesely
+//                2013  Johns Hopkins University (author: Daniel Povey)
+//                2014  Guoguo Chen
+
```
update copyright header.

danpovey commented on src/decoder/Makefile:
```diff
@@ -7,6 +7,7 @@ TESTFILES =
 
 OBJFILES = training-graph-compiler.o lattice-simple-decoder.o lattice-faster-decoder.o \
    lattice-faster-online-decoder.o simple-decoder.o faster-decoder.o \
+	 lattice-incremental-decoder.o \
```
again, odd whitespace.  These should be spaces not tabs.  Maybe you have your editor set up wrong.

danpovey commented on src/decoder/decoder-wrappers.cc:
```diff
@@ -286,6 +379,7 @@ bool DecodeUtteranceLatticeFaster(
       fst::ScaleLattice(fst::AcousticLatticeScale(1.0 / acoustic_scale), &lat);
     lattice_writer->Write(utt, lat);
   }
+  KALDI_VLOG(2) << "Delay time after decoding finalized (secs): " << timer.Elapsed();
```
obviously for the vesion we check in, this type of thing won't be there.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,602 @@
+                   "delay (in frames) at "
+                   "which to incrementally determinize lattices");
+    opts->Register("determinize-max-active", &determinize_max_active,
+                   "This option is to adaptively decide --determinize-delay. "
```
This does not do a very good job of documenting the option.  You don't clarify, for example, whether you are
supposed to set both of them or just one, or if this overrides the other one.  You don't say what a typical value mught be.   Also, for me this option is a bit problematic because it introduces a lot of uncertainty ; you would not be confident that the delay or memory usage would stay reasonably bounded.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,602 @@
+    opts->Register("redeterminize", &redeterminize,
+                   "whether to re-determinize the lattice after incremental "
+                   "determinization.");
+    opts->Register("determinize-lattice", &determinize_lattice,
```
I'd say this option probably doesn't make much sense for this decoder because you wouldn't want to use this decoder if you didn't want  determinization.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,602 @@
+  }
+  void Check() const {
+    KALDI_ASSERT(beam > 0.0 && max_active > 1 && lattice_beam > 0.0 &&
+                 min_active <= max_active && prune_interval > 0 &&
```
You should make sure to check that your new configs have suitable values.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,602 @@
+template <typename FST>
+class LatticeIncrementalDeterminizer;
+
+/** This is the "normal" lattice-generating decoder.
```
It's very important to update this type of documentation.  This is just copied.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,602 @@
+  // of it as a `pre-final state`
+  // Similarly, we define a `initial arc` as an arc from a initial-state, and the
+  // destination state of it as a `post-initial state`
+  // The initial states are constructed corresponding to pre-final states in the
```
you mean post-initial states here, maybe?

danpovey reviewed (commented):
some more comments.  But I am working on an email about a better version of the algorithm, so we shouldn't focus on cosmetic stuff just yet.

chenzhehuai commented:
> some more comments. But I am working on an email about a better version of the algorithm, so we shouldn't focus on cosmetic stuff just yet.

@danpovey  I also scratch a better description of our algorithm here (section 3). 
https://www.overleaf.com/4634381751wyjcqgbykvrf
Hope it can relieve some of your efforts

chenzhehuai committed:
[experimental] new det algorithm (#31)

the new algorithm is to determinize "states in the appended lattice with final-arcs to also have non-final arcs leaving them"

Commit:
adding redet frames

Commit:
add eps removal; 1oco

Commit:
bug fix when --epsilon-removal=1 --redeterminize-max-frames=10

Commit:
code refine

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,652 @@
+                   "Interval (in frames) at "
+                   "which to prune tokens");
+    opts->Register("determinize-delay", &determinize_delay,
+                   "delay (in frames, typically larger than --prune-interval) "
```
Capitalize Delay.  It's not clear to me why it should be larger than prune-interval (anyway, in the default, it's not).
You should clarify that smaller values lead to less latency in the determinized output.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,652 @@
+    opts->Register("determinize-delay", &determinize_delay,
+                   "delay (in frames, typically larger than --prune-interval) "
+                   "at which to incrementally determinize lattices.");
+    opts->Register("determinize-max-active", &determinize_max_active,
```
In the version we check in, let's take this out.  I don't think the adaptive delay is necessary any more now that we resolved the issues around pruning-- we might not even need it for the paper.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,652 @@
+                   "--determinize-delay to further reduce the computation "
+                   "introduced by incremental determinization. ");
+    opts->Register("redeterminize", &redeterminize,
+                   "whether to re-determinize the lattice after incremental "
```
I think we can leave this option out.   If people want fully-determinized lattices they can leave redeterminize-max-frames at infinity.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,652 @@
+    opts->Register("redeterminize", &redeterminize,
+                   "whether to re-determinize the lattice after incremental "
+                   "determinization.");
+    opts->Register("redeterminize_max_frames", &redeterminize_max_frames,
```
I'd add `Smaller -> less deterministic but less likely to blow up in bad cases.`

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,652 @@
+                   "redeterminize states.  This is mainly intended to avoid "
+                   "pathological cases. You could set it infinite to get a fully "
+                   "determinized lattice.");
+    opts->Register("epsilon-removal", &epsilon_removal,
```
In the version we check in, this shouldn't be an option.  It is probably not even needed for the paper, if you've already debugged that code.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,652 @@
+  bool GetBestPath(Lattice *ofst, bool use_final_probs = true);
+
+  // The following function is specifically designed for incremental
+  // determinization. The function obtains a CompactLattice for
```
Since this comment is so long I'd use `/** */` comment syntax.  The double asterisk is necessary to enable
doxygen, which is what the `@param` stuff is for (the alternative is triple slash, `///`).

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,652 @@
+  //                         in the lattice.  Otherwise treat all final-costs of
+  //                         states active
+  //                         on the most recent frame as zero (i.e.  Weight::One()).
+  //   @param [in]  redeterminize    If true, re-determinize the CompactLattice
```
You can remove this option now that our code gives fully determinized output.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,652 @@
+  //                         This does extra work, but not nearly as much as
+  //                         determinizing
+  //                          a RawLattice from scratch.
+  //   @param [in]  last_frame_of_chunk  Pass the last frame of this chunk to
```
Hm, so it looks like the prune-interval and determinize-delay are actually used outside this class?
If so, the config class should clearly document this, e.g. in a comment.
Make sure this works correctly if it's called twice on the same frame, we want it robust to user behavior.

danpovey commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,652 @@
+  /// Specifically design when decoding_finalized_==true
+  bool GetLattice(CompactLattice *olat);
+
+  /// This function is to keep forwards compatibility.
```
I think you mean backwards compatibility.

danpovey replied:
Actually I think you can delete this function.  If people wanted the raw lattice they could just use a different decoder.

chenzhehuai commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1618 @@
+      // numbers from frame_det_most to frame_det_least
+      for (int32 f = frame_det_most; f >= frame_det_least; f--) {
+        if (config_.determinize_max_active == std::numeric_limits<int32>::max() ||
+            GetNumToksForFrame(f) < config_.determinize_max_active) {
```
config_.determinize_max_active is used here

danpovey commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1618 @@
+    // We always incrementally determinize the lattice after lattice pruning in
+    // PruneActiveTokens()
+    // We have a delay on GetLattice to do determinization on more skinny lattices
+    int32 frame_det_most = NumFramesDecoded() - config_.determinize_delay;
```
This Decode() function is really just to show how it's run.  It wouldn't make much to use this decoder at all in a situation where you were calling Decode(), because you could just determinize at the end.  In reality people would be expected to call AdvanceDecoding() and then manually call GetLattice().  So the logic for deciding when we get the lattice would be driven by the user, most likely based on constraints from the application itself (e.g. the size of packets the client sends).  So I don't think it makes sense to put any complicated logic into here.  We just need to demonstrate how the interface works, that's all.

chenzhehuai replied:
Do you mean remove Decode() and implement AdvanceDecoding() without tightly integrating GetLattice() ? 
I was wondering whether we need to also give an example on how to call AdvanceDecoding() and GetLattice() with config_.determinize_delay, and where we should put this example

danpovey commented:
I think we can still have a Decode(), but make it clear that it's only
intended as an example, and keep it simple.
I think Decode() itself can probably be the example.  We may not even need
to check in the binary, once you have finished testing.  You can also
probably remove the decoder wrappers for the final PR, since I can't think
of a situation where we would need this functionality and yet it would
still be necessary to use the decoder wrappers.  But you should probably
keep them in a separate branch for testing purposes, for now at lesat.



On Sun, May 5, 2019 at 11:53 PM Zhehuai Chen <notifications@github.com>
wrote:

> *@chenzhehuai* commented on this pull request.
> ------------------------------
>
> In src/decoder/lattice-incremental-decoder.cc
> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r281057233>:
>
> > +    DecodableInterface *decodable) {
> +  InitDecoding();
> +
> +  // We use 1-based indexing for frames in this decoder (if you view it in
> +  // terms of features), but note that the decodable object uses zero-based
> +  // numbering, which we have to correct for when we call it.
> +
> +  while (!decodable->IsLastFrame(NumFramesDecoded() - 1)) {
> +    if (NumFramesDecoded() % config_.prune_interval == 0) {
> +      PruneActiveTokens(config_.lattice_beam * config_.prune_scale);
> +    }
> +
> +    // We always incrementally determinize the lattice after lattice pruning in
> +    // PruneActiveTokens()
> +    // We have a delay on GetLattice to do determinization on more skinny lattices
> +    int32 frame_det_most = NumFramesDecoded() - config_.determinize_delay;
>
> Do you mean remove Decode() and implement AdvanceDecoding() without
> tightly integrating GetLattice() ?
> I was wondering whether we need to also give an example on how to call
> AdvanceDecoding() and GetLattice() with config_.determinize_delay, and
> where we should put this example
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r281057233>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZFLO2OPQPJDW5OUQDVAQDPT6TS3ANCNFSM4HBXOKMA>
> .
>

danpovey commented:
Don't worry about removing that stuff just yet though.

On Mon, May 6, 2019 at 12:00 AM Daniel Povey <dpovey@gmail.com> wrote:

> I think we can still have a Decode(), but make it clear that it's only
> intended as an example, and keep it simple.
> I think Decode() itself can probably be the example.  We may not even need
> to check in the binary, once you have finished testing.  You can also
> probably remove the decoder wrappers for the final PR, since I can't think
> of a situation where we would need this functionality and yet it would
> still be necessary to use the decoder wrappers.  But you should probably
> keep them in a separate branch for testing purposes, for now at lesat.
>
>
>
> On Sun, May 5, 2019 at 11:53 PM Zhehuai Chen <notifications@github.com>
> wrote:
>
>> *@chenzhehuai* commented on this pull request.
>> ------------------------------
>>
>> In src/decoder/lattice-incremental-decoder.cc
>> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r281057233>:
>>
>> > +    DecodableInterface *decodable) {
>> +  InitDecoding();
>> +
>> +  // We use 1-based indexing for frames in this decoder (if you view it in
>> +  // terms of features), but note that the decodable object uses zero-based
>> +  // numbering, which we have to correct for when we call it.
>> +
>> +  while (!decodable->IsLastFrame(NumFramesDecoded() - 1)) {
>> +    if (NumFramesDecoded() % config_.prune_interval == 0) {
>> +      PruneActiveTokens(config_.lattice_beam * config_.prune_scale);
>> +    }
>> +
>> +    // We always incrementally determinize the lattice after lattice pruning in
>> +    // PruneActiveTokens()
>> +    // We have a delay on GetLattice to do determinization on more skinny lattices
>> +    int32 frame_det_most = NumFramesDecoded() - config_.determinize_delay;
>>
>> Do you mean remove Decode() and implement AdvanceDecoding() without
>> tightly integrating GetLattice() ?
>> I was wondering whether we need to also give an example on how to call
>> AdvanceDecoding() and GetLattice() with config_.determinize_delay, and
>> where we should put this example
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r281057233>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AAZFLO2OPQPJDW5OUQDVAQDPT6TS3ANCNFSM4HBXOKMA>
>> .
>>
>

Commit:
bug fix
add warning
add determinize_chunk_size

danpovey commented:
@chenzhehuai if this is close to being ready do you want to make a PR to Kaldi's github and we can have Hainan review it?

Commit:
code refine

Commit:
We need to be careful about the case where the start state of the
`appended lattice' becomes a redeterminized-state.
Certain things break down then, we might get an empty lattice.  This
could happen if there is silence at the utterance start.
It would likely be fixable by inserting an epsilon transition from a new
start state to the old start state, doing the normal procedure,
      and then removing that epsilon (the same way we would likely do
          when applying the `max-redeterminize-frames` option).

hainan-xv commented on src/bin/latgen-incremental-mapped.cc:
```diff
@@ -0,0 +1,177 @@
+    bool determinize = true;
+    CompactLatticeWriter compact_lattice_writer;
+    LatticeWriter lattice_writer;
+    if (! (determinize ? compact_lattice_writer.Open(lattice_wspecifier)
```
No space for the unary operator ! according to Google style guide.

hainan-xv commented on src/bin/latgen-incremental-mapped.cc:
```diff
@@ -0,0 +1,177 @@
+
+        for (; !loglike_reader.Done(); loglike_reader.Next()) {
+          std::string utt = loglike_reader.Key();
+          Matrix<BaseFloat> loglikes (loglike_reader.Value());
```
I feel this might copy the whole matrix to a new memory location? Maybe you could use Swap() to avoid the mem operation?

hainan-xv commented on src/bin/latgen-incremental-mapped.cc:
```diff
@@ -0,0 +1,177 @@
+            tot_like += like;
+            frame_count += loglikes.NumRows();
+            num_success++;
+          } else num_fail++;
```
Add brackets for else (since there are for the if block)

hainan-xv commented on src/bin/latgen-incremental-mapped.cc:
```diff
@@ -0,0 +1,177 @@
+          tot_like += like;
+          frame_count += loglikes.NumRows();
+          num_success++;
+        } else num_fail++;
```
brackets

hainan-xv commented on src/bin/latgen-incremental-mapped.cc:
```diff
@@ -0,0 +1,177 @@
+    double elapsed = timer.Elapsed();
+    KALDI_LOG << "Time taken "<< elapsed
+              << "s: real-time factor assuming 100 frames/sec is "
+              << (elapsed*100.0/frame_count);
```
parenthesis not required; space around /

hainan-xv commented on src/bin/latgen-incremental-mapped.cc:
```diff
@@ -0,0 +1,177 @@
+    if (num_success != 0) return 0;
+    else return 1;
+  } catch(const std::exception &e) {
+    std::cerr << e.what();
```
Probably better to use KALDI_ERR?

danpovey commented:
That std::cerr thing is standard, actually.  it's to show things like
allocation errors.

On Tue, May 28, 2019 at 3:16 PM hainan-xv <notifications@github.com> wrote:

> *@hainan-xv* commented on this pull request.
> ------------------------------
>
> In src/bin/latgen-incremental-mapped.cc
> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r288261436>:
>
> > +    }
> +
> +    double elapsed = timer.Elapsed();
> +    KALDI_LOG << "Time taken "<< elapsed
> +              << "s: real-time factor assuming 100 frames/sec is "
> +              << (elapsed*100.0/frame_count);
> +    KALDI_LOG << "Done " << num_success << " utterances, failed for "
> +              << num_fail;
> +    KALDI_LOG << "Overall log-likelihood per frame is " << (tot_like/frame_count) << " over "
> +              << frame_count<<" frames.";
> +
> +    delete word_syms;
> +    if (num_success != 0) return 0;
> +    else return 1;
> +  } catch(const std::exception &e) {
> +    std::cerr << e.what();
>
> Probably better to use KALDI_ERR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chenzhehuai/kaldi/pull/28?email_source=notifications&email_token=AAZFLOYTMPE22W46HDO5JL3PXWAJ5A5CNFSM4HBXOKMKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZ46RNA#pullrequestreview-242870452>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZFLO4UKQ5DELBY3MJEI6LPXWAJ5ANCNFSM4HBXOKMA>
> .
>

hainan-xv commented on src/decoder/decoder-wrappers.cc:
```diff
@@ -195,6 +195,87 @@ DecodeUtteranceLatticeFasterClass::~DecodeUtteranceLatticeFasterClass() {
 
+template <typename FST>
+bool DecodeUtteranceLatticeIncremental(
+    LatticeIncrementalDecoderTpl<FST> &decoder, // not const but is really an input.
```
In general passing non-const reference is not allowed in Google style -- if it gets changed, then pass a pointer and move it to the end of the args list.

hainan-xv commented on src/decoder/decoder-wrappers.cc:
```diff
@@ -195,6 +195,87 @@ DecodeUtteranceLatticeFasterClass::~DecodeUtteranceLatticeFasterClass() {
+    double *like_ptr) { // puts utterance's like in like_ptr on success.
+  using fst::VectorFst;
+  if (!decoder.Decode(&decodable)) {
+    KALDI_WARN << "Failed to decode file " << utt;
```
Probably should be "Failed to decode utterance with id"

danpovey commented:
I'm OK with the const thing to stay, as it's 'semantically' const, and this
is part of a pattern.
Regarding the message: that change makes sense, but change the same message
when it
appears elsewhere.


On Tue, May 28, 2019 at 3:20 PM hainan-xv <notifications@github.com> wrote:

> *@hainan-xv* commented on this pull request.
> ------------------------------
>
> In src/decoder/decoder-wrappers.cc
> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r288262975>:
>
> > +    LatticeIncrementalDecoderTpl<FST> &decoder, // not const but is really an input.
> +    DecodableInterface &decodable, // not const but is really an input.
> +    const TransitionModel &trans_model,
> +    const fst::SymbolTable *word_syms,
> +    std::string utt,
> +    double acoustic_scale,
> +    bool determinize,
> +    bool allow_partial,
> +    Int32VectorWriter *alignment_writer,
> +    Int32VectorWriter *words_writer,
> +    CompactLatticeWriter *compact_lattice_writer,
> +    LatticeWriter *lattice_writer,
> +    double *like_ptr) { // puts utterance's like in like_ptr on success.
> +  using fst::VectorFst;
> +  if (!decoder.Decode(&decodable)) {
> +    KALDI_WARN << "Failed to decode file " << utt;
>
> Probably should be "Failed to decode utterance with id"
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chenzhehuai/kaldi/pull/28?email_source=notifications&email_token=AAZFLO7WNUTD5PHQ3ETH7LDPXWAZVA5CNFSM4HBXOKMKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZ467QY#pullrequestreview-242872259>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZFLO4UGWXXI6KUYVMXKLDPXWAZVANCNFSM4HBXOKMA>
> .
>

hainan-xv commented on src/decoder/decoder-wrappers.cc:
```diff
@@ -195,6 +195,87 @@ DecodeUtteranceLatticeFasterClass::~DecodeUtteranceLatticeFasterClass() {
+        std::string s = word_syms->Find(words[i]);
+        if (s == "")
+          KALDI_ERR << "Word-id " << words[i] << " not in symbol table.";
+        std::cerr << s << ' ';
```
Is there a reason that std::cerr and KALDI_ERR are used here in different messages?

danpovey commented:
It's always been like that in the decoders, so it will just say
utterance-id  word1 word2...
We might not have done it like that if we had done it from scratch, but the
pattern is fixed now.


On Tue, May 28, 2019 at 3:25 PM hainan-xv <notifications@github.com> wrote:

> *@hainan-xv* commented on this pull request.
> ------------------------------
>
> In src/decoder/decoder-wrappers.cc
> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r288264746>:
>
> > +    std::vector<int32> alignment;
> +    std::vector<int32> words;
> +    GetLinearSymbolSequence(decoded, &alignment, &words, &weight);
> +    num_frames = alignment.size();
> +    KALDI_ASSERT(num_frames == decoder.NumFramesDecoded());
> +    if (words_writer->IsOpen())
> +      words_writer->Write(utt, words);
> +    if (alignment_writer->IsOpen())
> +      alignment_writer->Write(utt, alignment);
> +    if (word_syms != NULL) {
> +      std::cerr << utt << ' ';
> +      for (size_t i = 0; i < words.size(); i++) {
> +        std::string s = word_syms->Find(words[i]);
> +        if (s == "")
> +          KALDI_ERR << "Word-id " << words[i] << " not in symbol table.";
> +        std::cerr << s << ' ';
>
> Is there a reason that std::cerr and KALDI_ERR are used here in different
> messages?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chenzhehuai/kaldi/pull/28?email_source=notifications&email_token=AAZFLO3L5UK2AJQKASNFYXTPXWBL3A5CNFSM4HBXOKMKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZ47QPI#pullrequestreview-242874429>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZFLO5WMGXX77KZUFTRB4TPXWBL3ANCNFSM4HBXOKMA>
> .
>

hainan-xv commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,644 @@
+        beam_delta(0.5),
+        hash_ratio(2.0),
+        prune_scale(0.1),
+        max_word_id(1e7) {}
```
I'm wondering if 1e7 is really unattainably big enough...? The a-billion-word dataset has close to 1 million vocab-size, which is only an order of magnitude smaller than this...

chenzhehuai replied:
I modify it to 1e8 instead of std::numeric_limits<int32>::max(). The reason here to have this config is because in our algorithm we need to make state labels and token labels outside of the range of word ids.

danpovey commented:
Could use std::numeric_limits<int32>::max()

On Tue, May 28, 2019 at 3:43 PM hainan-xv <notifications@github.com> wrote:

> *@hainan-xv* commented on this pull request.
> ------------------------------
>
> In src/decoder/lattice-incremental-decoder.h
> <https://github.com/chenzhehuai/kaldi/pull/28#discussion_r288271234>:
>
> > +  fst::DeterminizeLatticePhonePrunedOptions det_opts;
> +
> +  LatticeIncrementalDecoderConfig()
> +      : beam(16.0),
> +        max_active(std::numeric_limits<int32>::max()),
> +        min_active(200),
> +        lattice_beam(10.0),
> +        prune_interval(25),
> +        determinize_delay(25),
> +        determinize_chunk_size(20),
> +        determinize_max_active(std::numeric_limits<int32>::max()),
> +        redeterminize_max_frames(std::numeric_limits<int32>::max()),
> +        beam_delta(0.5),
> +        hash_ratio(2.0),
> +        prune_scale(0.1),
> +        max_word_id(1e7) {}
>
> I'm wondering if 1e7 is really unattainably big enough...? The
> a-billion-word dataset has close to 1 million vocab-size, which is only an
> order of magnitude smaller than this...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chenzhehuai/kaldi/pull/28?email_source=notifications&email_token=AAZFLO4HZNMF2XUDVZKBQLLPXWDO5A5CNFSM4HBXOKMKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZ5BPKI#pullrequestreview-242882473>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZFLO4AT4AZL3LFLWKXZH3PXWDO5ANCNFSM4HBXOKMA>
> .
>

hainan-xv commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,644 @@
+   will normally be StdToken, but also may be BackpointerToken which is to support
+   quick lookup of the current best path (see lattice-faster-online-decoder.h)
+
+   The FST you invoke this decoder with is expected to equal
```
by equal you mean "of type"?

hainan-xv commented on src/decoder/lattice-incremental-decoder.h:
```diff
@@ -0,0 +1,644 @@
+  }
+
+  /// Outputs an FST corresponding to the single best path through the lattice.
+  /// Returns true if result is nonempty (using the return status is deprecated,
```
I'm guessing this is copied from other decoder code? Why not just make it void now since it's been decided as deprecated.

chenzhehuai replied:
I still implement this function by GetLattice() at first and then ShortestPath() from the compact lattice. Do you think we should remove this function?

hainan-xv replied:
Hmm I think you meant you are using the return value here when implementing those 2 functions. I guess in that case we should keep it and change the comments then?

@danpovey is the reason that this bool return value was going to be taken out because that we could just test whether *ofst is empty? I guess we could simply use a comma expression in the if statements here and make the function void?

hainan-xv reviewed (commented):
Submitting current comments while in the process of understanding the core decoder code.

Commit:
code refine

Commit:
Do the following modification. Results can be referred to sheet "ver 8" of https://docs.google.com/spreadsheets/d/1M_AlG1SYnDN663Fh8U0WMu7JrwKO9I_H5rCFXj6soI0/edit?usp=sharing

Let a {\em final-arc} be an arc with an arc with a {\em state-label} on it.  Suppose some state $s$ in the already-determinized part of the lattice has at least on {\em final-arc} leaving it, and it also has at least one successor-state $r$ whose time is more than {\em redeterminize-frames} earlier than the last frame in the already-determinized part of the lattice and which has no {\em final-arc} leaving it.  Before processing the already-determinized part of the lattice to determine which states are {\em redeterminized states} as described in Section~\ref{sec:determinizing_second_half}*, we modify it as follows: for each such state $s$: we add a new state $t$, put an epsilon transition from $s$ to $t$, and move all the arcs which are {\em not} to successor-states such as state $r$ above, to leave from $t$ instead of $s$.  This modification ensures that states such as $r$ do not become {\em redeterminized-states}, i.e. it limits the number of states we have to redeterminize.  We will remove the newly-introduced epsilons after determinizing the latest piece of the lattice and combining it with the already-determinized portion.

Commit:
make terms consistent with the paper

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+    // of final arcs to denote the "future" information of final states (Tokens)
+    // Moreover, the delay on GetLattice to do determinization
+    // make it process more skinny lattices which reduces the computation overheads.
+    int32 frame_det_most = NumFramesDecoded() - config_.determinize_delay;
```
this variable name (and also the *least one)is a bit confusing when I first read it. Maybe a better name would be something like determinize_window_right/left?

chenzhehuai replied:
Here they are used to decide the right boundary of the window/chunk for determinization. Hence I feel using determinize_window_right/left seems confusing

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+
+  We can use the extra_cost to accurately prune away tokens that we know will
+  never appear in the lattice.  If the extra_cost is greater than the desired
+  lattice beam, the token would provably never appear in the lattice, so we can
```
probably

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+                                     last_frame_of_chunk);
+    last_get_lattice_frame_ = last_frame_of_chunk;
+  } else if (last_get_lattice_frame_ > last_frame_of_chunk)
+    KALDI_WARN << "Call GetLattice up to frame: " << last_frame_of_chunk
```
there needs to be a bracket for the else if part because of there is one for the if part before (consistency according to Google style)

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+
+  ofst->DeleteStates();
+  unordered_multimap<int, StateId>
+      token_label2last_state_map; // for GetInitialRawLattice
```
this variable name, and also the tok_map/tmp_vec later -- I think I was told that it's better to not have words like map/vec in variable names. Not sure if it's really a rule, but here token_label2last_state_map, _map is redundant; the tok_map though isn't descriptive enough; maybe change it to token2state_id?

chenzhehuai replied:
I change them to token_label2last_state

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+               "e.g. the lattice reaches --max-mem constrain.";
+        continue;
+      }
+      std::vector<BaseFloat> tmp_vec;
```
this variable is never used

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+
+  GetRedeterminizedStates();
+
+  olat->DeleteStates();
```
It seems to me that when this function is called, olat should already be cleared, at least in the use case provided in this code. 

But maybe this is OK -- I guess this function could be called by the user.

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+      if (range.first == range.second) {
+        KALDI_WARN
+            << "The token in the first frame of this chunk does not "
+               "exist in the last frame of previous chunk. It should be seldom"
```
It should seldom happen and would be caused by...

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+        // for the arcs outgoing from the last frame Token in this chunk, we will
+        // create these arcs in the next chunk
+        if (f == frame_end && l->ilabel > 0) continue;
+        typename unordered_map<Token *, StateId>::const_iterator iter =
```
Why do we need typename here?

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+}
+
+template <typename FST, typename Token>
+BaseFloat LatticeIncrementalDecoderTpl<FST, Token>::ProcessEmitting(
```
it seems this function, and many other functions in this class, are identical to existing decoder code. I was wondering if it makes sense to change the inheritance structures of the classes so there could be a cleaner way to share the code?

chenzhehuai replied:
@danpovey what do you think about that?

danpovey replied:
I'd rather keep them separate.  That way there's less chance of breaking other decoders.

hainan-xv reviewed (commented):
some more comments.

Commit:
code refine according to Hainan's comments

hainan-xv commented on src/decoder/lattice-incremental-decoder.cc:
```diff
@@ -0,0 +1,1617 @@
+      // If the number of tokens in a certain frame is less than
+      // config_.determinize_max_active, the lattice can be determinized up to this
+      // frame. And we try to determinize as most frames as possible so we check
+      // numbers from frame_det_most to frame_det_least
```
It's probably better to add some comments here mentioning that no matter what happened during the search for chunks that meet the max_active threshold, all of the remaining chunks will be processed during the FinalizeDecoding() function later.

Commit:
add final-prune-after-determinize

Commit:
more comments

Commit:
bug fix

Commit:
refine

Commit:
add online decoder

Commit:
refine

Commit:
refine

Commit:
fix the bug to introduce one more eps arc sometime
